docs/RFCS/cluster_upgrade_tool.md

```diff
@@ -0,0 +1,107 @@
+- Feature Name: Cluster Upgrade Tool
+- Status: draft
+- Start Date: 2016-09-15
+- Authors: Daniel Harrison
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+A series of hooks to perform any necessary bookkeeping before a CockroachDB
+version upgrade.
+
+# Motivation
+
+Backward compatibility is crucial to a database, but can introduce considerable
+complexity to a codebase. All else being equal, it's better to keep backward
+compatibility logic siloed from the significant complexity that is inherent in
+software as complex as CockroachDB.
+
+# Detailed design
+
+Some versions of CockroachDB will require that a "system migration hook" be run
+before the cluster is upgraded to that version. This is expected to be
+infrequent; not every release will bump the version, only ones that need a
+migration hook.
+
+Migration hooks consist of a name, a work function, and optionally a minimum
+version. In the steady state, there is a map in code between migration names and
```

Comment:
I'm not sure that we need this map; I doubt we'll ever have enough migrations here that it's worthwhile to roll them up into a single value. Maintaining this map seems like it would be as much of a headache as giving all the migrations a version number instead of a name.